### PR TITLE
feat(highlights)!: redesign of highlighting approach

### LINKED ISSUE #1440: bug: hop.nvim doesn't work correctly when the text contains Korean, Japanese, and Chinese
### Did you check docs and existing issues?

- [X] I have read all the vscode-neovim docs
- [X] I have searched the existing issues of vscode-neovim
- [X] I can reproduce the issue with an empty init.vim/init.lua and no other vscode extensions (when applicable)

### Neovim version (nvim -v)

0.9.1

### Operating system/version

MacOS 13.4.1

### Describe the bug

https://github.com/vscode-neovim/vscode-neovim/assets/29355285/bba4071f-9f2d-438a-90db-81545b6a6a67

In alphabets like Korean, Japanese, and Chinese, hop.nvim's jump characters don't display correctly and strangely overlap with existing characters.

I tested this in Neovim with the same text and it worked perfectly. This error only appears in vscode-neovim.

I using this version of hop.nvim: [smoka7/hop.nvim: Neovim motions on speed!](https://github.com/smoka7/hop.nvim)

And I've already tested this things as well: [phaazon/hop.nvim: Neovim motions on speed!](https://github.com/phaazon/hop.nvim)

### Steps To Reproduce

1. type Korean or Japanese or Chinese
2. :HopWord

### Expected Behavior

The jump characters should appear right above the existing text, just like in English.

### PR DESCRIPTION
**Pull request by xiyaowong**
This PR rewrites the approach for handling highlights. It now renders cells based on columns. It aims to maintain consistency with nvim in most daily use cases. Only a few rare edge cases in daily work may still have issues.

- rendering
Fixes #1440

- syntax
Resolve: #519 by https://github.com/xiyaowong/vscode-neovim/pull/1
Close: #1468
Close: #799

Some demonstration images.

- hop & flash

- tabs

- emojis

emojis

- simple benchmarks

Regular document
benchmark1

Unusual document 👀 
benchmark2

Although there is noticeable lag in the second image, I still think it is sufficient.

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -52,14 +52,7 @@ module.exports = {
...
                 ],
-                "@typescript-eslint/explicit-function-return-type": [
-                    "warn",
-                    {
-                        allowExpressions: true,
-                        allowTypedFunctionExpressions: true,
-                        allowHigherOrderFunctions: true,
-                    },
-                ],
+                "@typescript-eslint/explicit-function-return-type": "off",
```
Comment (xiyaowong) [inline .eslintrc.js:55]: This rule is quite annoying.

**Comment by 5h5ong** [comment]
https://github.com/vscode-neovim/vscode-neovim/assets/29355285/fc292fd3-e488-4cc5-9d8e-5a339a35f17e

https://github.com/vscode-neovim/vscode-neovim/assets/29355285/871d7dcf-b5ab-4c5c-8b75-5b85161c9efa

https://github.com/vscode-neovim/vscode-neovim/assets/29355285/54e95f2e-8f39-4233-8656-62b96176ed1b

https://github.com/vscode-neovim/vscode-neovim/assets/29355285/3d6306aa-16dc-4e3b-9663-f5ec571ff3fa

I'm using `flash.nvim`. It works smoothly in real usage environments. It seems that the problems i used to face have been wll resolved.

**Comment by xiyaowong** [comment]
Marked as “ready for review” to get attention and assistance with testing from others.

**Comment by theol0403** [comment]
I'm extremely busy these days so might take me a bit to review this, just because it's a sensitive area and there are a lot of changes. Thanks for your work on this! 
Fwiw eol highlighting is important for fake visual cursor.

**Comment by theol0403** [comment]
Also, I haven't really looked at the code, but I feel like the highlight provider has the potential to reduce a lot of lines of code/simplify, if that's at all something you can work towards, as it was pretty messy before.

**Comment by xiyaowong** [comment]
This PR took up almost all of my time last week 😷. For the most part, it’s stable in most scenarios now, and I’m currently using the latest compiled extensions in my daily work. The remaining issue lies with multi-byte characters due to differences in handling character encoding length between positions in vim, the editor, and CSS.

**Comment by theol0403** [comment]
Hahah it's easy to get sucked into spending a lot of time on this project 😉

**Comment by xiyaowong** [comment]
But these are mostly edge cases that most people wouldn’t encounter in their daily work. For example, if you try selecting, moving, or manipulating this string: “🎉🔧🥇🎉 🔧 🥇 🤝 🎉 🔧 🥇 ❤️”, you’ll see where the issue lies. Through debugging, I’ve confirmed that the current position calculations are problematic for multi-byte characters. While I attempted to fix it, I found that it would impact other modules and introduce new issues. Therefore, I decided not to continue with the fix as I don’t want to invest too much effort in solving a rarely encountered problem. Personally, I prefer addressing such rare cases in the future.

**Comment by theol0403** [comment]
A quick test, everything seems to be working. I'll have to review the code later.

**Comment by theol0403** [comment]
Sometimes, when using leap.nvim, I get some labels that float & get stuck well past the end of the line:
image

**Comment by justinmk** [comment]
> Sometimes ... some labels float & get stuck

I see that with the current release (CTRL-L clears it up), so I wouldn't say it's a blocker for this PR.

**Comment by xiyaowong** [comment]
@theol0403 @justinmk  Solution for syntax. https://github.com/xiyaowong/vscode-neovim/pull/1

**Comment by xiyaowong** [comment]
@theol0403 I think this pull request is stable now. It can be reviewed and considered for merging. Due to the numerous changes in this PR, it has become challenging to proceed with other refactoring tasks.

**Review by theol0403 — APPROVED** [review]
Sorry for the wait, it looks great. Just clean up a few of the commented code that isn't being used. 

I don't have time to extensively test, but we can fix things as they come up.

**Comment by theol0403** [comment]
Feel free to merge after you've done the final touches.

**Comment by stevenmqnguyen** [comment]
Given this redesign, is flash.nvim able to do away with [manually setting colors](https://github.com/folke/flash.nvim/commit/d4c30b169f01b8108c5bc38e230a975408133603) and instead use the colors of current vscode theme? 

I'm not sure if I completely understand how the interaction between vscode-neovim and flash.nvim work but I tested removing the custom vscode logic in flash.nvim and it seems like in vscode the highlights aren't showing up.

**Comment by xiyaowong** [comment]
@stevenmqnguyen 

> Given this redesign, is flash.nvim able to do away with manually setting

No, plugins should not link to the builtin groups. 

There are two ways to customize colors: 
1. set colors in nvim
2. set colors in vscode. 
	References: 
	- https://github.com/vscode-neovim/vscode-neovim/blob/4d7a9be9be95ea5e8da793fd810345171a865cd2/package.json#L190C1-L200C26
	- [ThemeColor](https://code.visualstudio.com/api/references/theme-color)

**Pull request merged**